crypto/bn/bn_dsa.c

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright 2019 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+#include <stdio.h>
+/* The line above prevents "ISO C forbids an empty translation unit" warning */
+#ifndef FIPS_MODE
+# include <string.h>
+# include <openssl/evp.h>
+# include <openssl/rand_drbg.h>
+# include "internal/cryptlib.h"
+# include "internal/bn_int.h"
+# include "../rand/rand_lcl.h"
```

Comment:
I assume this rand_lcl.h is needed for something?

Reply:
@slontis Yes, it is about setting min_lengths.
If this header is not included, I cannot access drbg->min_entropylen and drbg->min_noncelen in this file.

Reply:
Ideally this would be done by an API..

Reply:
@slontis Yes, so I added the TODO to remind us to rewrite the code if it can be done by an API

Reply:
I think this PR should *strongly consider* adding the API's that are needed and not increase the technical debt.  Yes, that is a pain.  They can be internal API's.

Reply:
@richsalz Do you know what is the use of the marco `RAND_DRBG_STRENGTH`
https://github.com/openssl/openssl/blob/master/include/openssl/rand_drbg.h#L44-L65
How will you take it into account in the future?

Reply:
I believe it's an internal-only, pinging the real expert @mspncp

Reply:
> @richsalz Do you know what is the use of the marco `RAND_DRBG_STRENGTH`
> https://github.com/openssl/openssl/blob/2a1e2fe145c6eb8e75aa2e1b3a8c3a49384b2852/include/openssl/rand_drbg.h#L44-L65
> How will you take it into account in the future?
> If this is taken in to account, then there is no need to add a new API.

Hmmm... when this constant was introduced in c16de9d8329d41a2433d0f273c080d9d06ad7a87, it served a simple but important purpose: It was the **default security strength** of the DRBG on which the implementation of the default RAND method, RAND_OpenSSL() was based.  At this early stage, only the CTR_DRBG was ported from the FIPS 2.0 module and its security strength was not configurable yet at runtime. This is the reason why a preprocessor constant was used.

Here is the original #definition (in an internal header, because the RAND_DRBG interface was not public yet):

https://github.com/openssl/openssl/blob/c16de9d8329d41a2433d0f273c080d9d06ad7a87/include/internal/rand.h#L16-L25

Since then, a lot has changed, in particular the other DRBG variants were added (by @slontis) and some APIs were added to configure them at runtime. It seems to me that through those changes, the `RAND_DRBG_STRENGTH` constant has lost its original importance and is practically useless today. At least that is my impression when I look at the grep hits in the current master:

```
~/openssl$ git grep -nH RAND_DRBG_STRENGTH
crypto/rand/rand_lcl.h:79:                                  3 * (RAND_DRBG_STRENGTH / 16))
crypto/rand/rand_lcl.h:82: *                                1.5 * (RAND_DRBG_STRENGTH / 8))
crypto/rand/rand_lib.c:334:        pool = rand_pool_new(RAND_DRBG_STRENGTH,
crypto/rand/rand_lib.c:335:                             (RAND_DRBG_STRENGTH + 7) / 8,
crypto/rand/rand_unix.c:267:     *   2 * RAND_POOL_FACTOR * (RAND_DRBG_STRENGTH / 8) = 2^14
crypto/rand/randfile.c:80:     * of 'RAND_DRBG_STRENGTH' bytes (not bits!). This is done on purpose
crypto/rand/randfile.c:85:#define RAND_LOAD_BUF_SIZE (RAND_BUF_SIZE + RAND_DRBG_STRENGTH)
crypto/rand/randfile.c:115:            bytes = RAND_DRBG_STRENGTH;
include/openssl/rand_drbg.h:49: * does not take RAND_DRBG_STRENGTH into account and sets the strength of the
include/openssl/rand_drbg.h:52: * RAND_DRBG_STRENGTH is currently only used for the legacy RAND
include/openssl/rand_drbg.h:61:# define RAND_DRBG_STRENGTH             256
```

So even though the above comment states

```
* RAND_DRBG_STRENGTH is currently only used for the legacy RAND
* implementation.
```


it looks to me that the `RAND_DRBG_STRENGTH` constant has been superseded and only a few remnants of it remain, which are practically useless. Probably those last remnants should be examined and removed. (Not in this pr, of course.)

But let's return to your original problem: it looks like a hack to me to tweak internal parameters of the DRBG, which are normally specified by the NIST standard.

https://github.com/openssl/openssl/blob/55effe82ae23d8bef23594e0aaaaea20a197234a/crypto/bn/bn_dsa.c#L114-L116

For that reason, it does not make much sense IMHO to have API functions (accessors) for those values (neither public nor private). I understand that your case is special, because you instantiate the HMAC-DRBG with non-random 'entropy'. But still, there must be a better and more consistent way to do this. I'll take a look at the RFC and try if I can come up with an idea. But it will take a little time.


P.S: @JemmyLoveJenny I like your pull request, you are doing a great job!  :-)

Reply:
@mspncp Do you have any idea now?
What is the proper way to set drbg_hmac min_lengths in this special case?